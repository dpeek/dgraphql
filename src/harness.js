import fs from 'fs'
import path from 'path'

import { graphql } from 'graphql'

import { buildSchema } from './schema'

import type { ClientConfig } from './client'

export async function init (config: ClientConfig) {
  const schemaPath = path.resolve(__dirname, '__tests__/schema.graphql')
  const sourcePath = path.resolve(__dirname, '__tests__/source.graphql')

  const source = fs.readFileSync(sourcePath).toString()

  const time = String(new Date().getTime() - 1495660000000)
  let commonVariables = { time }

  // can't use the same schema as some tests are in relay mode, the query is not
  const dataSchema = buildSchema(fs.readFileSync(schemaPath).toString(), {})
  const data = await graphql({
    schema: dataSchema,
    source,
    variableValues: commonVariables,
    contextValue: { language: 'en' }
  })
  const walk = node => {
    if (node.id && node.name) {
      commonVariables[node.name.split(' ')[0].toLowerCase()] = node.id
    }
    Object.values(node).forEach(value => {
      if (typeof value === 'object') walk(value)
    })
  }
  walk(data)

  const schema = buildSchema(fs.readFileSync(schemaPath).toString(), config)
  return (source, variables, language) => {
    return graphql({
      schema,
      source,
      variableValues: { ...commonVariables, ...(variables || {}) },
      contextValue: { language: language || 'en' }
    })
  }
}
