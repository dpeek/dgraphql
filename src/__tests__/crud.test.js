import { init } from './harness'

var sequence

beforeAll(async () => {
  const test = await init()
  sequence = test.sequence
})

test('creates node', () => {
  const query1 = `mutation Test {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `query Test ($tim: ID!) {
    person(id: $tim) {
      name
    }
  }`
  return sequence([query1, query2])
})

test('returns fields for created node', () => {
  const query = `mutation Test {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        name
      }
    }
  }`
  return sequence([query])
})

test('updates node', () => {
  const query1 = `mutation Test {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!) {
    updatePerson(input:{id: $tim, name: "Jim"}) {
      person {
        name
      }
    }
  }`
  const query3 = `query Test ($tim: ID!) {
    person(id: $tim) {
      name
    }
  }`
  return sequence([query1, query2, query3])
})

test('returns fields on updated node', () => {
  const query1 = `mutation Test {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!) {
    updatePerson(input:{id: $tim, name: "Jim"}) {
      person {
        name
      }
    }
  }`
  return sequence([query1, query2])
})

test('deletes node', () => {
  const query1 = `mutation Test {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!) {
    deletePerson(input: {id: $tim}) {
      person {
        id
      }
    }
  }`
  const query3 = `query Test ($tim: ID!) {
    person(id: $tim) {
      name
    }
  }`
  return sequence([query1, query2, query3])
})

test('deleting non-existent node returns error', () => {
  const query = `mutation Test {
    deletePerson(input: {id: "0x00"}) {
      person {
        id
      }
    }
  }`
  return sequence([query])
})

test('deletes reverse edge to deleted node', () => {
  const query1 = `mutation Test {
    createPerson(input: {
      name: "Tim",
      partner: {
        name: "Bob"
      },
      emergencyContact: {
        name: "Harry"
      }
    }) {
      person {
        id
        name
        partner {
          id
          name
        }
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!) {
    deletePerson(input: {id: $tim}) {
      person {
        name
      }
    }
  }`
  const query3 = `query Test ($bob: ID!) {
    person(id: $bob) {
      partner {
        name
      }
    }
  }`
  return sequence([query1, query2, query3])
})

test('deletes reverse edges to deleted node', () => {
  const query1 = `mutation Test {
    createPerson(input: {
      name: "Tim",
      children: [{
        name: "Sarah"
      }, {
        name: "James"
      }]
    }) {
      person {
        id
        name
        children(order: name_asc) {
          id
          name
        }
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!) {
    deletePerson(input: {id: $tim}) {
      person {
        name
        children(order: name_asc) {
          name
        }
      }
    }
  }`
  const query3 = `query Test ($sarah: ID!) {
    person(id: $sarah) {
      name
      parents {
        name
      }
    }
  }`
  return sequence([query1, query2, query3])
})

test('creates node with nested node', () => {
  const query1 = `mutation Test {
    createPerson(input: {
      name: "Tim",
      partner: {
        name: "Bob"
      }
    }) {
      person {
        id
        name
        partner {
          id
          name
        }
      }
    }
  }`
  const query2 = `query Test ($tim: ID!, $bob: ID!) {
    partner1: person(id: $tim) {
      name
    }
    partner2: person(id: $bob) {
      name
    }
  }`
  return sequence([query1, query2])
})

test('creates node linked to existing nodes', () => {
  const query1 = `mutation Test {
    mum: createPerson(input: { name: "Linda" }) {
      person {
        id
        name
      }
    }
    dad: createPerson(input: { name: "Matthew" }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($linda: ID!, $matthew: ID!) {
    createPerson(
      input: {
        name: "David",
        parents:[
          { id: $linda },
          { id: $matthew }
        ]
      }
    ) {
      person {
        name
        parents(order: name_asc) {
          name
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('creating node with edge to non-existent node returns error', () => {
  const query1 = `mutation Test {
    createPerson(
      input: {
        name: "David",
        partner: { id: "0x00" }
      }
    ) {
      person {
        name
        partner {
          name
        }
      }
    }
  }`
  return sequence([query1])
})

test('creating node with edge to node of incorrect type returns error', () => {
  const query1 = `mutation Test {
    createPerson(input: { name: "Tim" }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!) {
    createPerson(
      input: {
        name: "David",
        emails: { id: $tim }
      }
    ) {
      person {
        name
        emails {
          address
        }
      }
    }
  }`
  return sequence([query1, query2])
})
