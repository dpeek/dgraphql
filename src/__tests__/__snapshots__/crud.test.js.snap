// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates node 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`creates node 2`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "Tim",
    },
  },
}
`;

exports[`creates node linked to existing nodes 1`] = `
Object {
  "data": Object {
    "dad": Object {
      "person": Object {
        "name": "Matthew",
      },
    },
    "mum": Object {
      "person": Object {
        "name": "Linda",
      },
    },
  },
}
`;

exports[`creates node linked to existing nodes 2`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "David",
        "parents": Object {
          "0": Object {
            "name": "Linda",
          },
          "1": Object {
            "name": "Matthew",
          },
        },
      },
    },
  },
}
`;

exports[`creates node with nested node 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "Tim",
        "partner": Object {
          "name": "Bob",
        },
      },
    },
  },
}
`;

exports[`creates node with nested node 2`] = `
Object {
  "data": Object {
    "partner1": Object {
      "name": "Tim",
    },
    "partner2": Object {
      "name": "Bob",
    },
  },
}
`;

exports[`creating node with edge to node of incorrect type returns error 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`creating node with edge to node of incorrect type returns error 2`] = `
Object {
  "data": Object {
    "createPerson": null,
  },
  "errors": Object {
    "0": Object {
      "locations": Object {
        "0": Object {
          "column": 5,
          "line": 2,
        },
      },
      "message": "Cannot create edge \\"emails\\" from \\"Person\\" to \\"Person\\", should be \\"Email\\"",
      "path": Object {
        "0": "createPerson",
      },
    },
  },
}
`;

exports[`creating node with edge to non-existent node returns error 1`] = `
Object {
  "data": Object {
    "createPerson": null,
  },
  "errors": Object {
    "0": Object {
      "locations": Object {
        "0": Object {
          "column": 5,
          "line": 2,
        },
      },
      "message": "There is no \\"Person\\" with id \\"0x00\\"",
      "path": Object {
        "0": "createPerson",
      },
    },
  },
}
`;

exports[`deletes node 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`deletes node 2`] = `
Object {
  "data": Object {
    "deletePerson": Object {
      "person": Object {},
    },
  },
}
`;

exports[`deletes node 3`] = `
Object {
  "data": Object {
    "person": null,
  },
  "errors": Object {
    "0": Object {
      "locations": Object {
        "0": Object {
          "column": 5,
          "line": 2,
        },
      },
      "message": "There is no 'Node' with id '0x00'",
      "path": Object {
        "0": "person",
      },
    },
  },
}
`;

exports[`deletes reverse edge to deleted node 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "Tim",
        "partner": Object {
          "name": "Bob",
        },
      },
    },
  },
}
`;

exports[`deletes reverse edge to deleted node 2`] = `
Object {
  "data": Object {
    "deletePerson": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`deletes reverse edge to deleted node 3`] = `
Object {
  "data": Object {
    "person": Object {
      "partner": null,
    },
  },
}
`;

exports[`deletes reverse edges to deleted node 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "children": Object {
          "0": Object {
            "name": "James",
          },
          "1": Object {
            "name": "Sarah",
          },
        },
        "name": "Tim",
      },
    },
  },
}
`;

exports[`deletes reverse edges to deleted node 2`] = `
Object {
  "data": Object {
    "deletePerson": Object {
      "person": Object {
        "children": Object {
          "0": Object {
            "name": "James",
          },
          "1": Object {
            "name": "Sarah",
          },
        },
        "name": "Tim",
      },
    },
  },
}
`;

exports[`deletes reverse edges to deleted node 3`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "Sarah",
      "parents": Object {},
    },
  },
}
`;

exports[`deleting non-existent node returns error 1`] = `
Object {
  "data": Object {
    "deletePerson": null,
  },
  "errors": Object {
    "0": Object {
      "locations": Object {
        "0": Object {
          "column": 5,
          "line": 2,
        },
      },
      "message": "There is no 'Person' with id '0x00'",
      "path": Object {
        "0": "deletePerson",
      },
    },
  },
}
`;

exports[`returns fields for created node 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`returns fields on updated node 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`returns fields on updated node 2`] = `
Object {
  "data": Object {
    "updatePerson": Object {
      "person": Object {
        "name": "Jim",
      },
    },
  },
}
`;

exports[`updates node 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`updates node 2`] = `
Object {
  "data": Object {
    "updatePerson": Object {
      "person": Object {
        "name": "Jim",
      },
    },
  },
}
`;

exports[`updates node 3`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "Jim",
    },
  },
}
`;
